SetItemListenerOnReconnectTest.testListenersTerminateCluster_nonSmart_multipleServer

http://jenkins.hazelcast.com/view/Official%20Builds/job/Hazelcast-3.x-ZuluJDK13/26/testReport/com.hazelcast.client.listeners/SetItemListenerOnReconnectTest/testListenersTerminateCluster_nonSmart_multipleServer/

**Stacktrace**
```
java.lang.AssertionError: Cluster size is not correct. Expected: 3, actual: 2, instance index: 0
	at org.junit.Assert.fail(Assert.java:88)
	at com.hazelcast.test.HazelcastTestSupport.assertClusterSize(HazelcastTestSupport.java:1138)
	at com.hazelcast.test.HazelcastTestSupport$11.run(HazelcastTestSupport.java:1166)
	at com.hazelcast.test.HazelcastTestSupport.assertTrueEventually(HazelcastTestSupport.java:1351)
	at com.hazelcast.test.HazelcastTestSupport.assertTrueEventually(HazelcastTestSupport.java:1368)
	at com.hazelcast.test.HazelcastTestSupport.assertClusterSizeEventually(HazelcastTestSupport.java:1163)
	at com.hazelcast.test.HazelcastTestSupport.assertClusterSizeEventually(HazelcastTestSupport.java:1151)
	at com.hazelcast.client.listeners.AbstractListenersOnReconnectTest.testListenersTerminateCluster(AbstractListenersOnReconnectTest.java:288)
	at com.hazelcast.client.listeners.AbstractListenersOnReconnectTest.testListenersTerminateCluster_nonSmart_multipleServer(AbstractListenersOnReconnectTest.java:250)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:567)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:114)
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:106)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.lang.Thread.run(Thread.java:830)
```

```
Hiccups measured while running test 'testListenersTerminateCluster_nonSmart_multipleServer(com.hazelcast.client.listeners.SetItemListenerOnReconnectTest):'
09:27:45, accumulated pauses: 47 ms, max pause: 10 ms, pauses over 1000 ms: 0
09:27:50, accumulated pauses: 39 ms, max pause: 3 ms, pauses over 1000 ms: 0
09:27:55, accumulated pauses: 39 ms, max pause: 1 ms, pauses over 1000 ms: 0
09:28:00, accumulated pauses: 41 ms, max pause: 0 ms, pauses over 1000 ms: 0
09:28:05, accumulated pauses: 40 ms, max pause: 0 ms, pauses over 1000 ms: 0
09:28:10, accumulated pauses: 44 ms, max pause: 2 ms, pauses over 1000 ms: 0
09:28:15, accumulated pauses: 45 ms, max pause: 1 ms, pauses over 1000 ms: 0
09:28:20, accumulated pauses: 39 ms, max pause: 0 ms, pauses over 1000 ms: 0
09:28:25, accumulated pauses: 39 ms, max pause: 0 ms, pauses over 1000 ms: 0
09:28:30, accumulated pauses: 40 ms, max pause: 0 ms, pauses over 1000 ms: 0
09:28:35, accumulated pauses: 38 ms, max pause: 0 ms, pauses over 1000 ms: 0
09:28:40, accumulated pauses: 38 ms, max pause: 0 ms, pauses over 1000 ms: 0
09:28:45, accumulated pauses: 37 ms, max pause: 0 ms, pauses over 1000 ms: 0
09:28:50, accumulated pauses: 38 ms, max pause: 1 ms, pauses over 1000 ms: 0
09:28:55, accumulated pauses: 39 ms, max pause: 0 ms, pauses over 1000 ms: 0
09:29:00, accumulated pauses: 33 ms, max pause: 0 ms, pauses over 1000 ms: 0
09:29:05, accumulated pauses: 40 ms, max pause: 5 ms, pauses over 1000 ms: 0
09:29:10, accumulated pauses: 34 ms, max pause: 0 ms, pauses over 1000 ms: 0
09:29:15, accumulated pauses: 38 ms, max pause: 1 ms, pauses over 1000 ms: 0
09:29:20, accumulated pauses: 34 ms, max pause: 0 ms, pauses over 1000 ms: 0
09:29:25, accumulated pauses: 36 ms, max pause: 0 ms, pauses over 1000 ms: 0
09:29:30, accumulated pauses: 37 ms, max pause: 0 ms, pauses over 1000 ms: 0
09:29:35, accumulated pauses: 37 ms, max pause: 0 ms, pauses over 1000 ms: 0
09:29:40, accumulated pauses: 38 ms, max pause: 1 ms, pauses over 1000 ms: 0
09:29:45, accumulated pauses: 35 ms, max pause: 0 ms, pauses over 1000 ms: 0
09:29:50, accumulated pauses: 1215 ms, max pause: 1204 ms, pauses over 1000 ms: 1
```

--- a/hazelcast/src/main/java/com/hazelcast/client/impl/connection/nio/ClientConnectionManagerImpl.java
+++ b/hazelcast/src/main/java/com/hazelcast/client/impl/connection/nio/ClientConnectionManagerImpl.java
@@ -348,7 +348,6 @@ public class ClientConnectionManagerImpl implements ClientConnectionManager {
             try {
                 doConnectToCluster();
                 synchronized (clientStateMutex) {
-
                     connectToClusterTaskSubmitted = false;
                     if (activeConnections.isEmpty()) {
                         if (logger.isFineEnabled()) {
@@ -419,7 +418,7 @@ public class ClientConnectionManagerImpl implements ClientConnectionManager {
             logger.warning("Exception during initial connection to " + address + ": " + e);
             throw e;
         } catch (Exception e) {
-            logger.warning("Exception during x initial connection to " + address + ": " + e);
+            logger.warning("Exception during initial connection to " + address + ": " + e);
             return null;
         }
     }
@@ -823,6 +822,15 @@ public class ClientConnectionManagerImpl implements ClientConnectionManager {
 
             fireConnectionAddedEvent(connection);
         }
+
+        // It could happen that this connection is already closed and
+        // onConnectionClose() is called even before the synchronized block
+        // above is executed. In this case, now we have a closed but registered
+        // connection. We do a final check here to remove this connection
+        // if needed.
+        if (!connection.isAlive()) {
+            onConnectionClose(connection);
+        }
     }
 
     private ClientMessage encodeAuthenticationRequest() {
@@ -928,7 +936,7 @@ public class ClientConnectionManagerImpl implements ClientConnectionManager {
             logger.warning("Failure during sending state to the cluster.", e);
             synchronized (clientStateMutex) {
                 if (targetClusterId.equals(clusterId)) {
-                    if (logger.isSevereEnabled()) {
+                    if (logger.isFineEnabled()) {
                         logger.warning("Retrying sending state to the cluster: " + targetClusterId + ", name: " + clusterName);
                     }
 

